From 430003a9d5a73f4a883b5e0466c84c1935d20b47 Mon Sep 17 00:00:00 2001
From: Minju Kim <minju2.kim@lge.com>
Date: Wed, 12 Jun 2019 18:13:23 +0900
Subject: [PATCH] fixup! [ose][acg][error_page] Register
 "com.webos.settingsservice.client"

:Release Notes:
Limit the permission only for network error page to launch "settings" app

:Detailed Notes:
Only network error page has the permission to launch "settings" app.
Before this patch, all the web apps can launch "settings" app with
"com.webos.settingsservice.client-*" luna service.

:Testing Performed:
Tested QO-3861

:QA Notes:
Test-cases: QO-3861

Depens on http://gpro.lge.com/242426

:Issues Addressed:
[PLAT-80157] Port "ACG related" patches from webosose/chromium68 -
Network settings button from network error page

Change-Id: I36a47bdb955e03aa019fc0151c23584a392bae73
---
 files/sysbus/com.palm.webappmanager.perm.json    | 2 +-
 files/sysbus/com.palm.webappmanager.role.json.in | 4 ++--
 src/core/WebPageBase.cpp                         | 7 +++++++
 src/core/WebPageBase.h                           | 1 +
 src/platform/webengine/BlinkWebView.cpp          | 7 +++++++
 src/platform/webengine/BlinkWebView.h            | 2 +-
 src/platform/webengine/WebPageBlink.cpp          | 4 ++++
 src/platform/webengine/WebPageBlink.h            | 1 +
 src/platform/webengine/WebPageBlinkDelegate.h    | 1 +
 9 files changed, 25 insertions(+), 4 deletions(-)

diff --git a/files/sysbus/com.palm.webappmanager.perm.json b/files/sysbus/com.palm.webappmanager.perm.json
index fad2918..fc01878 100644
--- a/files/sysbus/com.palm.webappmanager.perm.json
+++ b/files/sysbus/com.palm.webappmanager.perm.json
@@ -11,7 +11,7 @@
     "com.webos.rm.client.*": [
         "media"
     ],
-    "com.webos.settingsservice.client": [
+    "com.webos.settingsservice.client-*": [
         "applications.internal",
         "settings.read"
     ]
diff --git a/files/sysbus/com.palm.webappmanager.role.json.in b/files/sysbus/com.palm.webappmanager.role.json.in
index 1d81b93..81e8483 100644
--- a/files/sysbus/com.palm.webappmanager.role.json.in
+++ b/files/sysbus/com.palm.webappmanager.role.json.in
@@ -5,7 +5,7 @@
         \"com.palm.webappmanager\",
         \"com.webos.service.webappmanager\",
         \"com.webos.rm.client.*\",
-        \"com.webos.settingsservice.client\"
+        \"com.webos.settingsservice.client-*\"
     ],
     \"permissions\": [
       {
@@ -26,7 +26,7 @@
         \"outbound\":[\"com.webos.media\"]
       },
       {
-        \"service\":\"com.webos.settingsservice.client\",
+        \"service\":\"com.webos.settingsservice.client-*\",
         \"outbound\":[
           \"com.webos.applicationManager\",
           \"com.webos.settingsservice\"
diff --git a/src/core/WebPageBase.cpp b/src/core/WebPageBase.cpp
index 895c443..c0e1712 100644
--- a/src/core/WebPageBase.cpp
+++ b/src/core/WebPageBase.cpp
@@ -31,12 +31,15 @@
 
 #define CONSOLE_DEBUG(AAA) evaluateJavaScript(QStringLiteral("console.debug('") + QStringLiteral(AAA) + QStringLiteral("');"))
 
+const char kIdentifierForNetErrorPage[] = "com.webos.settingsservice.client";
+
 WebPageBase::WebPageBase()
     : m_appDesc(nullptr)
     , m_suspendAtLoad(false)
     , m_isClosing(false)
     , m_isLoadErrorPageFinish(false)
     , m_isLoadErrorPageStart(false)
+    , m_didErrorPageLoadedFromNetErrorHelper(false)
     , m_enableBackgroundRun(false)
     , m_loadErrorPolicy(QStringLiteral("default"))
     , m_cleaningResources(false)
@@ -51,6 +54,7 @@ WebPageBase::WebPageBase(const QUrl& url, std::shared_ptr<ApplicationDescription
     , m_isClosing(false)
     , m_isLoadErrorPageFinish(false)
     , m_isLoadErrorPageStart(false)
+    , m_didErrorPageLoadedFromNetErrorHelper(false)
     , m_enableBackgroundRun(false)
     , m_defaultUrl(url)
     , m_launchParams(params)
@@ -88,6 +92,8 @@ QString WebPageBase::getIdentifier() const
 
     if(appId().isEmpty() || appId() == WebAppManager::instance()->getContainerAppId())
         return QStringLiteral("");
+    if ((m_isLoadErrorPageFinish && m_isLoadErrorPageStart) || m_didErrorPageLoadedFromNetErrorHelper)
+        return QString(kIdentifierForNetErrorPage);
     return m_appId;
 }
 
@@ -247,6 +253,7 @@ void WebPageBase::urlChangedSlot()
 void WebPageBase::handleLoadStarted()
 {
     m_suspendAtLoad = true;
+    m_didErrorPageLoadedFromNetErrorHelper = false;
 }
 
 void WebPageBase::handleLoadFinished()
diff --git a/src/core/WebPageBase.h b/src/core/WebPageBase.h
index bf070b9..9f34e8b 100644
--- a/src/core/WebPageBase.h
+++ b/src/core/WebPageBase.h
@@ -198,6 +198,7 @@ protected:
     bool m_isClosing;
     bool m_isLoadErrorPageFinish;
     bool m_isLoadErrorPageStart;
+    bool m_didErrorPageLoadedFromNetErrorHelper;
     bool m_enableBackgroundRun;
     QUrl m_defaultUrl;
     QString m_launchParams;
diff --git a/src/platform/webengine/BlinkWebView.cpp b/src/platform/webengine/BlinkWebView.cpp
index 609e5d5..e73028c 100644
--- a/src/platform/webengine/BlinkWebView.cpp
+++ b/src/platform/webengine/BlinkWebView.cpp
@@ -259,3 +259,10 @@ void BlinkWebView::DidResumeDOM()
     if (m_delegate)
         m_delegate->didResumeDOM();
 }
+
+void BlinkWebView::DidErrorPageLoadedFromNetErrorHelper()
+{
+    if (!m_delegate)
+        return;
+    return m_delegate->didErrorPageLoadedFromNetErrorHelper();
+}
diff --git a/src/platform/webengine/BlinkWebView.h b/src/platform/webengine/BlinkWebView.h
index 39dac65..0baddeb 100644
--- a/src/platform/webengine/BlinkWebView.h
+++ b/src/platform/webengine/BlinkWebView.h
@@ -68,7 +68,7 @@ public:
     void HandleBrowserControlFunction(const std::string& command, const std::vector<std::string>& arguments, std::string* result) override;
     void LoadVisuallyCommitted() override;
     void DidResumeDOM() override;
-
+    void DidErrorPageLoadedFromNetErrorHelper() override;
 
 private:
     WebPageBlinkDelegate* m_delegate;
diff --git a/src/platform/webengine/WebPageBlink.cpp b/src/platform/webengine/WebPageBlink.cpp
index d5d8d8f..cbefb82 100644
--- a/src/platform/webengine/WebPageBlink.cpp
+++ b/src/platform/webengine/WebPageBlink.cpp
@@ -746,6 +746,10 @@ void WebPageBlink::loadFailed(const std::string& url, int errCode, const std::st
     handleLoadFailed(errCode);
 }
 
+void WebPageBlink::didErrorPageLoadedFromNetErrorHelper() {
+   m_didErrorPageLoadedFromNetErrorHelper = true;
+}
+
 void WebPageBlink::loadVisuallyCommitted()
 {
     m_hasBeenShown = true;
diff --git a/src/platform/webengine/WebPageBlink.h b/src/platform/webengine/WebPageBlink.h
index 5f04680..1609242 100644
--- a/src/platform/webengine/WebPageBlink.h
+++ b/src/platform/webengine/WebPageBlink.h
@@ -121,6 +121,7 @@ public:
     void renderProcessCrashed() override;
     void titleChanged(const std::string& title) override;
     void navigationHistoryChanged() override;
+    void didErrorPageLoadedFromNetErrorHelper() override;
 
     void updateExtensionData(const QString& key, const QString& value);
     void setLoadErrorPolicy(const QString& policy);
diff --git a/src/platform/webengine/WebPageBlinkDelegate.h b/src/platform/webengine/WebPageBlinkDelegate.h
index 7f4e0e4..aa28d1e 100644
--- a/src/platform/webengine/WebPageBlinkDelegate.h
+++ b/src/platform/webengine/WebPageBlinkDelegate.h
@@ -51,6 +51,7 @@ public:
     virtual void didSwapCompositorFrame() {}
     virtual bool allowMouseOnOffEvent() const = 0;
     virtual void didResumeDOM() {}
+    virtual void didErrorPageLoadedFromNetErrorHelper() = 0;
 };
 
 #endif //WEBPAGE_BLINK_DELEGATE_H_
-- 
2.21.0

